### PR TITLE
Major cleanup and improvements around revision handling in the datastores

### PR DESCRIPTION
**Pull request by josephschorr**
1) Extract out the common revision types into a common library, add testing and change datastores to use the shared implementations
2) Change the non-decimal needed revisions to use a simple int64 or uint64
3) Change the HLC type used in CRDB to the CRDB-provided decimal type - this should result in reduced memory usage

### CODE REVIEW

File: internal/datastore/revisions/timestamprevision.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/authzed/spicedb/pkg/datastore"
+)
+
+// TimestampRevision is a revision that is a timestamp.
+type TimestampRevision int64
```
Comment (ecordell) [inline internal/datastore/revisions/timestamprevision.go:12]: why not uint64?

File: internal/datastore/revisions/hlcrevision.go [hunk 2]
```diff
@@ -3,19 +3,20 @@ package revisions
...
+	"github.com/cockroachdb/apd"
 	"github.com/shopspring/decimal"
 
 	"github.com/authzed/spicedb/pkg/datastore"
 )
 
 // HLCRevision is a revision that is a hybrid logical clock, stored as a decimal.
 type HLCRevision struct {
-	decimal decimal.Decimal
+	decimal *apd.Decimal
```
Comment (ecordell) [inline internal/datastore/revisions/hlcrevision.go:14]: why a pointer here? if we directly embed, then an array of hlcs gc just has to check 1 object isntead of all of the entires.

File: internal/datastore/revisions/timestamprevision.go (hunk 1 above)
Comment (josephschorr) [reply internal/datastore/revisions/timestamprevision.go:12]: Timestamps before supported negative numbers and I wasn't 100% certain if we needed to continue doing so, so I allowed it

File: internal/datastore/revisions/hlcrevision.go (hunk 2 above)
Comment (josephschorr) [reply internal/datastore/revisions/hlcrevision.go:14]: All the New* methods return a pointer and all the accessors on it are on the pointer type

Comment (josephschorr) [reply internal/datastore/revisions/hlcrevision.go]: Changed to grab the value and use that

**Review by ecordell — DISMISSED** [review]

**Review by ecordell — APPROVED** [review]

**Pull request merged**